# Домашнее задание 1 - PANDAS

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В каждом разделе вам предлагается выполнить несколько заданий. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотекам pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написаем production-кода.

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*.

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

### Формат сдачи
Для сдачи задания получившийся файл \*.ipynb с решением необходимо выложить в свой репозиторий github.

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных. С этой библиотекой мы сейчас и познакомимся.

Основной объект в pandas это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [96]:
%pylab inline
# import almost all we need
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Данные находятся в приложенном файле `chipotle.tsv`

#### 1. [0.5 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [97]:
file_path = 'chipotle.tsv'
df = pd.read_csv(file_path, sep='\t')
print(df.tail(10))

      order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25   
4613                                                NaN     $2.15   
4614                                                NaN     $1.50   
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans,

Данные представляют собой записи заказов, сделанных клиентами. Каждый заказ может включать несколько товаров. Столбцы содержат информацию о каждом товаре в заказе, включая его название, количество, цену и дополнительные опции или ингредиенты.

#### 2. [0.25 баллов] Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [98]:
items = df['item_name'].unique()
orders = df['order_id'].unique()
print(items)

['Chips and Fresh Tomato Salsa' 'Izze' 'Nantucket Nectar'
 'Chips and Tomatillo-Green Chili Salsa' 'Chicken Bowl' 'Side of Chips'
 'Steak Burrito' 'Steak Soft Tacos' 'Chips and Guacamole'
 'Chicken Crispy Tacos' 'Chicken Soft Tacos' 'Chicken Burrito'
 'Canned Soda' 'Barbacoa Burrito' 'Carnitas Burrito' 'Carnitas Bowl'
 'Bottled Water' 'Chips and Tomatillo Green Chili Salsa' 'Barbacoa Bowl'
 'Chips' 'Chicken Salad Bowl' 'Steak Bowl' 'Barbacoa Soft Tacos'
 'Veggie Burrito' 'Veggie Bowl' 'Steak Crispy Tacos'
 'Chips and Tomatillo Red Chili Salsa' 'Barbacoa Crispy Tacos'
 'Veggie Salad Bowl' 'Chips and Roasted Chili-Corn Salsa'
 'Chips and Roasted Chili Corn Salsa' 'Carnitas Soft Tacos'
 'Chicken Salad' 'Canned Soft Drink' 'Steak Salad Bowl'
 '6 Pack Soft Drink' 'Chips and Tomatillo-Red Chili Salsa' 'Bowl'
 'Burrito' 'Crispy Tacos' 'Carnitas Crispy Tacos' 'Steak Salad'
 'Chips and Mild Fresh Tomato Salsa' 'Veggie Soft Tacos'
 'Carnitas Salad Bowl' 'Barbacoa Salad Bowl' 'Salad' 'Veggie Cris

In [99]:
print(len(items))
print(orders)

50
[   1    2    3 ... 1832 1833 1834]


всего заказов: 1834
всего уникальных категорий товара: 50

#### 3. [0.25 баллов] Есть ли в данных пропуски? В каких колонках?

In [101]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
missing_values = df.isnull().sum() # Проверка на наличие пропущенных значений
print(missing_values)

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64


In [102]:
import pandas as pd

   # Загрузка файла с указанием разделителя
df = pd.read_csv('chipotle.tsv', sep='\t')
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip').dropna(how='all')


In [103]:
import pandas as pd
try:
    df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')

    missing_values = df.isnull().sum() # Проверка на наличие пропусков
    print(missing_values)
except Exception as e:
    print(f"Произошла ошибка: {e}")

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [74]:
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
df.fillna({col: '' for col in df.select_dtypes(include=['object']).columns}, inplace=True)
df.fillna(0, inplace=True)
print(df)

      order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                           $2.39   
1                                          [Clementine]  

#### 4. [0.5 баллов] Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [75]:
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
print(df['item_price'].dtype) # Проверка типа колонки item_price
def convert_price(price_str):
    return float(price_str.replace('$', '').strip())
df['price'] = df['item_price'].apply(convert_price) # Применение функции к item_price и создание колонки price
print(df[['item_price', 'price']].head())

object
  item_price  price
0     $2.39    2.39
1     $3.39    3.39
2     $3.39    3.39
3     $2.39    2.39
4    $16.98   16.98


Какая средняя/минимальная/максимальная цена у товара?

In [76]:
average_price = df['price'].mean()
min_price = df['price'].min()
max_price = df['price'].max()
print(f"Средняя цена: {average_price:.2f}")
print(f"Минимальная цена: {min_price:.2f}")
print(f"Максимальная цена: {max_price:.2f}")

Средняя цена: 7.46
Минимальная цена: 1.09
Максимальная цена: 44.25


Удалите старую колонку с ценой.

In [77]:
# Удаление старой колонки item_price
df = df.drop(columns=['item_price'])
print(df.head())

   order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  price  
0                                                NaN   2.39  
1                                       [Clementine]   3.39  
2                                            [Apple]   3.39  
3                                                NaN   2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...  16.98  


#### 5. [0.25 баллов] Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [78]:
unique_items = df[['item_name', 'price']].drop_duplicates() # Удаление дубликатов
sorted_items = unique_items.sort_values(by='price') # Сортировка по цене
# Получение 5 самых дешевых и 5 самых дорогих товаров
cheapest_items = sorted_items.head(5)
most_expensive_items = sorted_items.tail(5)
print("Самые дешевые товары:")
print(cheapest_items)
print("Самые дорогие товары:")
print(most_expensive_items)

Самые дешевые товары:
             item_name  price
34       Bottled Water   1.09
28         Canned Soda   1.09
263  Canned Soft Drink   1.25
329      Bottled Water   1.50
6        Side of Chips   1.69
Самые дорогие товары:
                         item_name  price
409                   Chicken Bowl  32.94
3601                Veggie Burrito  33.75
1254               Chicken Burrito  35.00
3480                 Carnitas Bowl  35.25
3598  Chips and Fresh Tomato Salsa  44.25


#### 6. [0.5 баллов] Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [79]:
chicken_bowl_orders = df[df['item_name'] == 'Chicken Bowl'].copy()
# Преобразование 'quantity' в целочисленный формат с использованием .loc
chicken_bowl_orders.loc[:, 'quantity'] = chicken_bowl_orders['quantity'].astype(int)
more_than_one_chicken_bowl = chicken_bowl_orders[chicken_bowl_orders['quantity'] > 1]
count = more_than_one_chicken_bowl.shape[0] # Получение общего количества таких заказов
print(f"Количество заказов, где клиенты покупали больше 1 Chicken Bowl: {count}")

Количество заказов, где клиенты покупали больше 1 Chicken Bowl: 33


#### 7. [0.5 баллов] Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [80]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
print("Имена столбцов:", df.columns) # Проверка имен столбцов
# Удаляем символы доллара, запятые и пробелы из 'item_price' при помощи str.replace()
df['item_price'] = df['item_price'].str.replace('$', '', regex=False)
df['item_price'] = df['item_price'].str.replace(',', '', regex=False)
df['item_price'] = df['item_price'].str.strip()  # Удаление лишних пробелов
df['item_price'] = df['item_price'].astype(float) # Преобразуем в float

grouped_orders = df.groupby('order_id').agg(total_price=('item_price', 'sum'),
                                            total_items=('quantity', 'sum')).reset_index()
average_check = grouped_orders['total_price'].mean()
average_items = grouped_orders['total_items'].mean()
print(f"Средний чек: {average_check:.2f} $")
print(f"Среднее количество товаров в заказе: {average_items:.2f}")

Имена столбцов: Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')
Средний чек: 18.81 $
Среднее количество товаров в заказе: 2.71


#### 8. [0.25 баллов] Сколько заказов содержали ровно 1 товар?

In [81]:
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
order_counts = df.groupby('order_id')['quantity'].sum().reset_index()
orders_with_one_item = order_counts[order_counts['quantity'] == 1]
num_orders_with_one_item = orders_with_one_item.shape[0]
print(f"Количество заказов, содержащих ровно 1 товар: {num_orders_with_one_item}")

Количество заказов, содержащих ровно 1 товар: 56


#### 9. [0.25 баллов] Какая самая популярная категория товара?

In [82]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
category_counts = df.groupby('item_name')['quantity'].sum().reset_index()
most_popular_item = category_counts.loc[category_counts['quantity'].idxmax()]
print(f"Самая популярная категория товара: {most_popular_item['item_name']} с количеством {most_popular_item['quantity']}")

Самая популярная категория товара: Chicken Bowl с количеством 761


#### 10. [0.5 баллов] Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой?

In [83]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t', on_bad_lines='skip')
burritos = df[df['item_name'].str.contains('Burrito', case=False)].copy()
burritos['item_price'] = burritos['item_price'].str.replace('$', '', regex=False).str.strip() # Очистка столбца item_price
burritos['item_price'] = burritos['item_price'].replace('', '0') # Замена пустых строк на '0'
burritos['item_price'] = burritos['item_price'].astype(float) # Преобразование в float
burrito_counts = burritos.groupby('item_name')['quantity'].sum().reset_index()
most_popular_burrito = burrito_counts.loc[burrito_counts['quantity'].idxmax()]
most_expensive_burrito = burritos.loc[burritos['item_price'].idxmax()]
print("Виды буррито:")
print(burrito_counts)
print(f"Самый популярный буррито: {most_popular_burrito['item_name']} с количеством {most_popular_burrito['quantity']}")
print(f"Самый дорогой буррито: {most_expensive_burrito['item_name']} с ценой ${most_expensive_burrito['item_price']:.2f}")

Виды буррито:
          item_name  quantity
0  Barbacoa Burrito        91
1           Burrito         6
2  Carnitas Burrito        60
3   Chicken Burrito       591
4     Steak Burrito       386
5    Veggie Burrito        97
Самый популярный буррито: Chicken Burrito с количеством 591
Самый дорогой буррито: Chicken Burrito с ценой $35.00


#### 11. [0.75 баллов] В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно среднюю стоимость заказа, добавить ее в исходные данные и сделать необходимые проверки.

*Данный комментарий стоит воспринимать как подсказку к одному из вариантов решений задания. Если в вашем варианте решения он не нужнен, это не страшно*

In [84]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t')

df['item_price'] = df['item_price'].astype(str) # Очистка столбца 'item_price' от символов '$', пробелов и запятых,для этого убедимся, что значения - строки

df['item_price'] = df['item_price'].str.replace('$', '', regex=False)
df['item_price'] = df['item_price'].str.replace(',', '', regex=False)
df['item_price'] = df['item_price'].str.strip()

df['item_price'] = pd.to_numeric(df['item_price'], errors='coerce') # Преобразуем в числовой тип (float)
order_summary = df.groupby('order_id').agg(
    total_price=('item_price', 'sum'),
    items=('item_name', 'count')
).reset_index()

df = df.merge(order_summary[['order_id', 'total_price']], on='order_id') # Объединяем информацию о заказах с исходными данными

df['is_expensive_item'] = df['item_price'] > (0.4 * df['total_price']) # Определяем, превышает ли цена товара 40% от суммы заказа

expensive_orders_count = df[df['is_expensive_item']]['order_id'].nunique() # Подсчитываем количество заказов, содержащих хотя бы один дорогой товар
print(f"Количество заказов с товаром, который стоит более 40% от суммы всего чека: {expensive_orders_count}")

Количество заказов с товаром, который стоит более 40% от суммы всего чека: 1624


#### 12. [0.75 баллов] Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [85]:
import pandas as pd
df = pd.read_csv('chipotle.tsv', sep='\t')
df['item_price'] = df['item_price'].astype(str)
df['item_price'] = df['item_price'].str.replace('$', '', regex=False)
df['item_price'] = df['item_price'].str.replace(',', '', regex=False)
df['item_price'] = df['item_price'].str.strip()
df['item_price'] = pd.to_numeric(df['item_price'], errors='coerce')  # kopiни NaN, если есть некорректные значения

initial_price = 1.25
new_price = 1.35

diet_coke_sales = df[df['item_name'] == 'Diet Coke']  # Фильтрация данных для Diet Coke

total_quantity = diet_coke_sales['quantity'].sum()

initial_income = total_quantity * initial_price # Рассчитываем доход до и после изменения цены
new_income = total_quantity * new_price

percentage_increase = ((new_income - initial_income) / initial_income) * 100
print(f"Общий доход от Diet Coke до изменения цены: ${initial_income:.2f}")
print(f"Общий доход от Diet Coke после изменения цены: ${new_income:.2f}")
print(f"Процент увеличения дохода: {percentage_increase:.2f}%")

Общий доход от Diet Coke до изменения цены: $0.00
Общий доход от Diet Coke после изменения цены: $0.00
Процент увеличения дохода: nan%


/tmp/ipython-input-2427532342.py:19: RuntimeWarning: invalid value encountered in scalar divide
  percentage_increase = ((new_income - initial_income) / initial_income) * 100


#### 13. [0.75 баллов] Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [86]:
import pandas as pd
data = [
    ['Убийственный класс', '(комедия/боевик)'],
    ['Стальной Алхимик: Братство', '(боевик)'],
    ['Саики Кусуо', '(комедия)'],
    ['Твоя апрельская ложь', '(романтика/драма)'],
    ['Нитидзё', '(комедия)'],
    ['Steins;Gate', '(фантастика/триллер)'],
    ['Хоримия', '(романтика)'],
    ['Моя геройская академия', '(комедия/боевик)'],
    ['Fate/Zero', '(триллер/боевик)'],
    ['Корзинка фруктов', '(романтика)'],
    ['Плутон', '(триллер/боевик)'],
    ['Галактика Татами', '(комедия/фантастика)'],
    ['Атака титанов', '(боевик/триллер)'],
    ['Ковбой Бибоп', '(фантастика/боевик)'],
    ['Шова Генроку Ракуго Синдзю', '(драма)']
]
df = pd.DataFrame(data, columns=['index', 'column1'])
df.set_index('index', inplace=True) # Установка первой колонки 'index' в качестве индекса
df.to_csv('chipotle.tsv', sep='t') # Сохранение DataFrame в файл chipotle.tsv с разделителем табуляция
print(df)

                                         column1
index                                           
Убийственный класс              (комедия/боевик)
Стальной Алхимик: Братство              (боевик)
Саики Кусуо                            (комедия)
Твоя апрельская ложь           (романтика/драма)
Нитидзё                                (комедия)
Steins;Gate                 (фантастика/триллер)
Хоримия                              (романтика)
Моя геройская академия          (комедия/боевик)
Fate/Zero                       (триллер/боевик)
Корзинка фруктов                     (романтика)
Плутон                          (триллер/боевик)
Галактика Татами            (комедия/фантастика)
Атака титанов                   (боевик/триллер)
Ковбой Бибоп                 (фантастика/боевик)
Шова Генроку Ракуго Синдзю               (драма)


Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [87]:
import pandas as pd

data = [
    ['Убийственный класс', '(комедия/боевик)'],
    ['Стальной Алхимик: Братство', '(боевик)'],
    ['Саики Кусуо', '(комедия)'],
    ['Твоя апрельская ложь', '(романтика/драма)'],
    ['Нитидзё', '(комедия)'],
    ['Steins;Gate', '(фантастика/триллер)'],
    ['Хоримия', '(романтика)'],
    ['Моя геройская академия', '(комедия/боевик)'],
    ['Fate/Zero', '(триллер/боевик)'],
    ['Корзинка фруктов', '(романтика)'],
    ['Плутон', '(триллер/боевик)'],
    ['Галактика Татами', '(комедия/фантастика)'],
    ['Атака титанов', '(боевик/триллер)'],
    ['Ковбой Бибоп', '(фантастика/боевик)'],
    ['Шова Генроку Ракуго Синдзю', '(драма)']
]
df = pd.DataFrame(data)
df.to_csv('chipotle.csv', index=False, header=False)
print(df)

                             0                     1
0           Убийственный класс      (комедия/боевик)
1   Стальной Алхимик: Братство              (боевик)
2                  Саики Кусуо             (комедия)
3         Твоя апрельская ложь     (романтика/драма)
4                      Нитидзё             (комедия)
5                  Steins;Gate  (фантастика/триллер)
6                      Хоримия           (романтика)
7       Моя геройская академия      (комедия/боевик)
8                    Fate/Zero      (триллер/боевик)
9             Корзинка фруктов           (романтика)
10                      Плутон      (триллер/боевик)
11            Галактика Татами  (комедия/фантастика)
12               Атака титанов      (боевик/триллер)
13                Ковбой Бибоп   (фантастика/боевик)
14  Шова Генроку Ракуго Синдзю               (драма)


## 2. Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать.

In [88]:
%matplotlib inline  # нужно для отображения графиков внутри ноутбука
import matplotlib.pyplot as plt

UsageError: unrecognized arguments: # нужно для отображения графиков внутри ноутбука


На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [89]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])

AttributeError: 'DataFrame' object has no attribute 'order_id'

#### 1. [1 балл] Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name.

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [90]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных
df = pd.read_csv('chipotle.tsv', sep='t')

# Создание нового столбца с ценами как суммы покупки (item_price)
# Он уже есть в данных, обычно, он в виде строки, убираем знак '$' и преобразуем к float
df['item_price_float'] = df['item_price'].replace('[$,]', '', regex=True).astype(float)

# Расчёт общей суммы покупок
# Общее количество покупок - это количество строк, для суммы - сумма по item_price_float
total_sales = df['item_price_float'].sum()

# Расчёт среднего значения цены по каждому виду товара (item_name)
mean_price_per_item = df.groupby('item_name')['item_price_float'].mean()

# Создание графиков
fig, axes = plt.subplots(1, 2, figsize=(15, 6))

# 1. Гистограмма распределения суммы покупок (общая сумма - можно показать, например, распределение по продаже отдельных позиций)
# Или, например, распределение цен отдельных товаров, чтобы было интересно
axes[0].hist(df['item_price_float'], bins=30, color='skyblue')
axes[0].set_title('Распределение цен товаров')
axes[0].set_xlabel('Цена ($)')
axes[0].set_ylabel('Количество')


# 2. Гистограмма распределения средних цен отдельных видов продуктов
axes[1].hist(mean_price_per_item, bins=20, color='salmon')
axes[1].set_title('Распределение средних цен по видам продуктов')
axes[1].set_xlabel('Средняя цена ($)')
axes[1].set_ylabel('Количество товаров')

plt.tight_layout()
plt.show()


KeyError: 'item_price'

In [46]:
import pandas as pd
import matplotlib.pyplot as plt
file_path = 'chipotle.tsv'
df = pd.read_csv('chipotle.tsv', sep='t')

print("Доступные колонки:", df.columns)

if 'item_price' in df.columns:
    df['item_price'] = df['item_price'].astype(str)
    df['item_price'] = df['item_price'].str.replace('$', '', regex=False)
    df['item_price'] = df['item_price'].str.replace(',', '', regex=False)
    df['item_price'] = df['item_price'].str.strip()
    df['item_price'] = pd.to_numeric(df['item_price'], errors='coerce')  # превращение в float

    # Проверим, что у нас есть данные после преобразования
    print(df['item_price'].describe())

    # Аггрегация данных по заказам
    order_summary = df.groupby('order_id').agg(
        total_price=('item_price', 'sum'),
        items=('item_name', 'count')
    ).reset_index()

    # Распределение суммы покупок по заказам
    purchase_sums = df.groupby('order_id')['item_price'].sum()

    # Средняя цена по продуктам
    average_prices = df.groupby('item_name')['item_price'].mean()

    # Построение гистограмм
    fig, axs = plt.subplots(1, 2, figsize=(12, 6))

    axs[0].hist(purchase_sums, bins=30, color='blue', alpha='0.7')
    axs[0].set_title('Распределение сумм покупок')
    axs[0].set_xlabel('Сумма покупок')
    axs[0].set_ylabel('Частота')

    axs[1].hist(average_prices, bins=30, color='green', alpha='0.7')
    axs[1].set_title('Распределение средних цен продуктов')
    axs[1].set_xlabel('Средняя цена')
    axs[1].set_ylabel('Частота')

    plt.tight_layout()
    plt.show()
else:
    print("Колонка 'item_price' не найдена. Доступные колонки:", df.columns)


Доступные колонки: Index(['index', 'column1'], dtype='object')
Колонка 'item_price' не найдена. Доступные колонки: Index(['index', 'column1'], dtype='object')


#### 2. [1 балл] Постройте график зависимости суммы покупок от дней.

In [45]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных из файла chipotle.tsv
df = pd.read_csv('chipotle.tsv', sep='\t')

# Преобразуем столбец 'item_price' в числовой формат (удаляем знак доллара и преобразуем в float)
df['item_price'] = df['item_price'].replace({'$': '', ',': ''}, regex=True).astype(float)

# Преобразуем столбец 'order_date' в формат datetime
df['order_date'] = pd.to_datetime(df['order_date'])

# Группируем данные по дате и суммируем суммы покупок
daily_sales = df.groupby(df['order_date'].dt.date)['item_price'].sum().reset_index()

# Переименовываем столбцы для удобства
daily_sales.columns = ['date', 'total_sales']

# Строим график
plt.figure(figsize=(12, 6))
plt.plot(daily_sales['date'], daily_sales['total_sales'], marker='o')
plt.title('Сумма покупок по дням')
plt.xlabel('Дата')
plt.ylabel('Сумма покупок ($)')
plt.xticks(rotation=45)
plt.grid()
plt.tight_layout()
plt.show()

KeyError: 'item_price'

In [39]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных из файла chipotle.tsv
df = pd.read_csv('chipotle.tsv', sep='\t')

df['item_price'] = df['item_price'].astype(str) # Очистка столбца 'item_price' от символов '$', пробелов и запятых,для этого убедимся, что значения - строки
df['item_price'] = df['item_price'].str.replace('$', '', regex=False)
df['item_price'] = df['item_price'].str.replace(',', '', regex=False)
df['item_price'] = df['item_price'].str.strip()
df['item_price'] = pd.to_numeric(df['item_price'], errors='coerce') # Преобразуем в числовой тип (float)
order_summary = df.groupby('order_id').agg(
    total_price=('item_price', 'sum'),
    items=('item_name', 'count')
).reset_index()



# Преобразуем столбец 'order_date' в datetime
df['order_date'] = pd.to_datetime(df['order_date'])

# Суммируем покупки по дням
daily_sales = df.groupby(df['order_date'].dt.date)['item_price'].sum()

# Создание графика
plt.figure(figsize=(12, 6))
plt.plot(daily_sales.index, daily_sales.values, marker='o', linestyle='-', color='blue')
plt.title('Сумма покупок по дням')
plt.xlabel('Дата')
plt.ylabel('Сумма покупок')
plt.xticks(rotation=45)
plt.grid()
plt.tight_layout()
plt.show()


KeyError: 'item_price'

#### 3. [1 балл] Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных из файла chipotle.tsv
df = pd.read_csv('chipotle.tsv', sep='\t')

# Предобработка данных
# Преобразуем столбец 'item_price' в числовой формат (удаляем знак доллара и преобразуем в float)
df['item_price'] = df['item_price'].replace({'$': '', ',': ''}, regex=True).astype(float)

# Преобразуем столбец 'order_date' в datetime
df['order_date'] = pd.to_datetime(df['order_date'])

# Добавляем столбец с днем недели (0 - понедельник, 6 - воскресенье)
df['day_of_week'] = df['order_date'].dt.day_name()

# Рассчитываем средние суммы покупок по дням недели
average_sales_by_day = df.groupby('day_of_week')['item_price'].mean()

# Упорядочиваем дни недели
days_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
average_sales_by_day = average_sales_by_day.reindex(days_order)

# Создание столбчатого графика
plt.figure(figsize=(10, 6))
average_sales_by_day.plot(kind='bar', color='skyblue')
plt.title('Средняя сумма покупок по дням недели')
plt.xlabel('День недели')
plt.ylabel('Средняя сумма покупок ($)')
plt.xticks(rotation=45)
plt.grid(axis='y')
plt.tight_layout()
plt.show()


FileNotFoundError: [Errno 2] No such file or directory: 'chipotle.tsv'

#### 4. [1 балл] Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных из файла chipotle.tsv
df = pd.read_csv('chipotle.tsv', sep='\t')

# Предобработка данных
# Преобразуем столбец 'item_price' в числовой формат (удаляем знак доллара и преобразуем в float)
df['item_price'] = df['item_price'].replace({'$': '', ',': ''}, regex=True).astype(float)

# Извлекаем количество купленных товаров из столбца 'quantity'
# Предполагаем, что количество находится в столбце 'quantity'
# Если его нет, то нужно будет его создать или извлечь из других данных
# В данном случае, предположим, что 'quantity' - это количество в каждой строке
# Если у вас есть другой способ получения количества, используйте его

# Создаем scatter plot
plt.figure(figsize=(10, 6))
plt.scatter(df['quantity'], df['item_price'], alpha=0.5, color='blue')
plt.title('Зависимость цены товара от количества')
plt.xlabel('Количество купленных товаров')
plt.ylabel('Цена за товар ($)')
plt.grid(True)
plt.tight_layout()
plt.show()


FileNotFoundError: [Errno 2] No such file or directory: 'chipotle.tsv'

Сохраните график в формате pdf (так он останется векторизованным).

In [31]:
import pandas as pd
import matplotlib.pyplot as plt

# Загрузка данных из файла chipotle.tsv
df = pd.read_csv('chipotle.tsv', sep='\t')

# Предобработка данных
# Преобразуем столбец 'item_price' в числовой формат (удаляем знак доллара и преобразуем в float)
df['item_price'] = df['item_price'].replace({'$': '', ',': ''}, regex=True).astype(float)

# Извлекаем количество купленных товаров из столбца 'quantity'
# Предполагаем, что количество находится в столбце 'quantity'
# Если его нет, то нужно будет его создать или извлечь из других данных
# В данном случае, предположим, что 'quantity' - это количество в каждой строке
# Если у вас есть другой способ получения количества, используйте его

# Создаем scatter plot
plt.figure(figsize=(10, 6))
plt.scatter(df['quantity'], df['item_price'], alpha=0.5, color='blue')
plt.title('Зависимость цены товара от количества')
plt.xlabel('Количество купленных товаров')
plt.ylabel('Цена за товар ($)')
plt.grid(True)
plt.tight_layout()

# Сохраняем график в формате PDF
plt.savefig('chipotle_price_quantity.pdf', format='pdf')

# Показываем график (опционально)
plt.show()


ValueError: could not convert string to float: '$2.39 '

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё.